pkg/pillar/zedcloud/authen.go

```diff
@@ -0,0 +1,508 @@
+	sm := &zcert.ZControllerCert{}
+	err := proto.Unmarshal(content, sm)
+	if err != nil {
+		log.Infof("SaveCertsToFile: Unmarshal err %v\n", err)
+		return err
+	}
+
+	var block *pem.Block
+	var leafcert *x509.Certificate
+	var certByte []byte
+	var keyCnt int
+	interm := x509.NewCertPool()
+	for _, cert := range sm.GetCerts() {
+		keyCnt++
+		if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_INTERMEDIATE {
+			ok := interm.AppendCertsFromPEM(cert.GetCert())
+			if !ok {
+				err := fmt.Errorf("VerifyCloudCertChain: fail to append intermediate certs")
+				return err
+			}
+			log.Infof("VerifyCloudCertChain: get intermediate cert %d, len %d\n", keyCnt, len(cert.GetCert()))
+		} else if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING {
+			if len(certByte) > 0 {
+				err := fmt.Errorf("VerifyCloudCertChain: received more than one leaf cert")
+				return err
+			}
+			certByte = cert.GetCert()
+			block, _ = pem.Decode(certByte)
+			if block == nil {
+				err := fmt.Errorf("VerifyCloudCertChain: can not get cert block")
+				return err
+			}
+			leafcert, err = x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				log.Infof("VerifyCloudCertChain: fail to parse certificate %v\n", err)
+				return err
+			}
+			log.Infof("VerifyCloudCertChain: get signing cert %d, len %d\n", keyCnt, len(cert.GetCert()))
+		} else {
+			err := fmt.Errorf("VerifyCloudCertChain: received cert with unknown type %d", cert.Type)
+			return err
+		}
+	}
+
+	log.Infof("VerifyCloudCertChain: key count %d\n", keyCnt)
+	if leafcert == nil {
+		err := fmt.Errorf("VerifyCloudCertChain: fail to acquire leaf cert")
+		return err
+	}
+
+	// Get the rootCAs from the zedcloudCtx
+	var roots *x509.CertPool
+	if ctx.TlsConfig != nil {
+		roots = ctx.TlsConfig.RootCAs
+	} else {
+		err := fmt.Errorf("VerifyCloudCertChain: no rootCAs in TlsConfig")
+		return err
+	}
+
+	opts := x509.VerifyOptions{
+		Roots:         roots,
+		DNSName:       serverName,
+		Intermediates: interm,
+	}
+	if _, err := leafcert.Verify(opts); err != nil {
+		log.Infof("VerifyCloudCertChain: fail to verify the cert chain %v\n", err)
+		return err
+	}
+
+	// save this cert and hash of it
+	cloudCertHash = ComputeSha(certByte)
+
+	err = saveServerCertToFile(certByte)
+	if err != nil {
+		log.Infof("VerifyCloudCertChain: file save err %v\n", err)
+		return err
+	}
+	log.Infof("VerifyCloudCertChain: success\n")
+	return nil
+}
+
+// saveCertsToFile - save cloud signing certs into files
+func saveServerCertToFile(certByte []byte) error {
+	tmpfileName := types.ServerCertName + ".tmp"
+	pemTmpCertFile, err := os.Create(tmpfileName)
+	if err != nil {
+		log.Infof("SaveCertsToFile: create tmp file err %v\n", err)
+		return err
+	}
+
+	_, err = pemTmpCertFile.Write(certByte)
+	if err != nil {
+		log.Infof("SaveCertsToFile: write pem to file err %v\n", err)
+		return err
+	}
+
+	defer os.Remove(pemTmpCertFile.Name())
```

Comment:
Move this defer up just after the Create block.

Reply:
ok.